File: backend/app/schemas/reservation.py
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field, ConfigDict
from app.models.allocation_request import AllocationRequestStatus
from app.models.reservation import ReservationStatus
from app.models.resource import ResourceType

class AllocationRequestCreate(BaseModel):
    hospital_id: int = Field(..., description="Target hospital ID for allocation request")

class AllocationRejectRequest(BaseModel):
    reason: str = Field(..., min_length=1, max_length=500, description="Rejection reason")

class AllocationRequestResponse(BaseModel):
    id: int
    emergency_case_id: int
    hospital_id: int
    status: AllocationRequestStatus
    match_score: float
    requested_at: datetime
    responded_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None
    rejection_reason: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)

class ReservationItem(BaseModel):
    resource_type: ResourceType
    quantity: int

class ReservationResponse(BaseModel):
    id: int
    emergency_case_id: int
    hospital_id: int
    resource_id: int
    quantity: int
    status: ReservationStatus
    reserved_at: datetime

    model_config = ConfigDict(from_attributes=True)

class AcceptSuccessResponse(BaseModel):
    request_id: int
    status: str = "ACCEPTED"
    hospital_id: int
    hospital_name: str
    reservations: List[ReservationItem]

File: frontend/src/types/reservation.ts
import type { ResourceType } from './hospital';

export type AllocationRequestStatus =
  | 'PENDING'
  | 'ACCEPTED'
  | 'REJECTED'
  | 'EXPIRED'
  | 'CANCELLED';

export type ReservationStatus =
  | 'HELD'
  | 'CONFIRMED'
  | 'RELEASED'
  | 'CONSUMED'
  | 'EXPIRED';

export interface AllocationRequest {
  id: number;
  emergency_case_id: number;
  hospital_id: number;
  status: AllocationRequestStatus;
  match_score: number;
  requested_at: string;
  responded_at?: string;
  expires_at?: string;
  rejection_reason?: string;
}

export interface ReservationItem {
  resource_type: ResourceType;
  quantity: number;
}

export interface AcceptSuccessResponse {
  request_id: number;
  status: string;
  hospital_id: number;
  hospital_name: string;
  reservations: ReservationItem[];
}

export interface Reservation {
  id: number;
  emergency_case_id: number;
  hospital_id: number;
  resource_id: number;
  quantity: number;
  status: ReservationStatus;
  reserved_at: string;
}
